components/tidb_query/src/expr_util/like.rs

```diff
@@ -11,38 +11,48 @@
 //!    in non-binary mode.
 
 use crate::expr::Result;
+use crate::expr_util::collation::{Charset, Collator};
 
-pub fn like(target: &[u8], pattern: &[u8], escape: u32) -> Result<bool> {
+pub fn like<C: Collator>(target: &[u8], pattern: &[u8], escape: u32) -> Result<bool> {
     // current search positions in pattern and target.
     let (mut px, mut tx) = (0, 0);
     // positions for backtrace.
     let (mut next_px, mut next_tx) = (0, 0);
     while px < pattern.len() || tx < target.len() {
-        if px < pattern.len() {
-            let c = pattern[px];
-            match c {
-                b'_' => {
-                    if tx < target.len() {
-                        px += 1;
-                        tx += 1;
-                        continue;
-                    }
-                }
-                b'%' => {
-                    // update the backtrace point.
-                    next_px = px;
-                    next_tx = tx + 1;
-                    px += 1;
+        if let Some((c, mut poff)) = C::Charset::decode_one(&pattern[px..]) {
+            let code: u32 = c.into();
+            if code == '_' as u32 {
+                if let Some((_, toff)) = C::Charset::decode_one(&target[tx..]) {
+                    px += poff;
+                    tx += toff;
                     continue;
                 }
-                mut pc => {
-                    if u32::from(pc) == escape && px + 1 < pattern.len() {
-                        px += 1;
-                        pc = pattern[px];
+            } else if code == '%' as u32 {
+                // update the backtrace point.
+                next_px = px;
+                px += poff;
+                next_tx = tx
+                    + if let Some((_, toff)) = C::Charset::decode_one(&target[tx..]) {
+                        toff
+                    } else {
+                        1
+                    };
+                continue;
+            } else {
+                if code == escape && px + 1 < pattern.len() {
```

Comment:
Do we still need `px + 1 < pattern.len()`?

Reply:
fided

Reply:
It's still necessary and should be fixed to `px + poff < pattern.len()`

Reply:
AH... I thought wrong.

Reply:
In MySQL:

```
mysql> select "C:\\" LIKE "C:\\";
+--------------------+
| "C:\\" LIKE "C:\\" |
+--------------------+
|                  0 |
+--------------------+
1 row in set (0.00 sec)
```
But the algorithm here gives `1`. The escape still has effect even if it's the last character.

Reply:
I add a testcase 

```rust
(r#"C:\\"#, r#"C:\\"#, '\\', Collation::Binary, Some(0)),
```

and it passed.

Reply:
It should be (r#"C:\"#, r#"C:\"#, '\\', Collation::Binary, Some(0)),

Reply:
And I test it on Mysql 5.7

```SQL
SELECT VERSION();
SELECT "C:\\" LIKE "C:\\";
```

```
5.7.28-0ubuntu0.18.04.4
1

```

Reply:
It seems that mysql will output 1 for this case?

Reply:
Anyway, it seems that the change is bug-compatible with current implementation both on TiDB and TiKV. It seems that it's better to file a new bug issue for this case in TiDB?

Reply:
I also tried:

```sql
SELECT VERSION();
select "x" LIKE "x" ESCAPE "x"
```

```
5.7.28-0ubuntu0.18.04.4
1
```

can't reproduce your result.

Reply:
also 8.0

Reply:
That's weird...

mysql> SELECT VERSION(); select 'x' LIKE 'x' ESCAPE "x";
+-----------+
| VERSION() |
+-----------+
| 5.7.28    |
+-----------+
1 row in set (0.00 sec)

+-------------------------+
| 'x' LIKE 'x' ESCAPE "x" |
+-------------------------+
|                       0 |
+-------------------------+
1 row in set (0.00 sec)

Reply:
It's true on MySQL 8.0. Let's keep the behavior here.